lib/reservation_manager.rb

```diff
@@ -0,0 +1,136 @@
+require_relative "reservation.rb"
+require "date"
+
+require "awesome_print"
+
+module Hotel
+  class ReservationManager
+    #add constant for room numbers
+    MAX_ROOMS = 20
+    ROOM_RATE = 200
+
+    def initialize
+      @reservations = []
+      @block = []
+    end
+
+    def all_rooms
+      rooms = (1..MAX_ROOMS).to_a
+      return rooms
+    end
+
+    def make_reservation(check_in, check_out)
+      # unsure how to raise argument error when room is unavailable since room_number is assigned from avail_rooms array
+      if available_rooms(check_in, check_out).length == 0
+        raise ArgumentError, "There are no available rooms for this time period"
+      else
+        room_number = available_rooms(check_in, check_out).first
+        # reservation_rate = reservation_cost(check_in, check_out)
+
+        new_reservation = Hotel::Reservation.new(room_number,
+                                                 check_in,
+                                                 check_out)
+      end
+      @reservations << new_reservation
+
+      return @reservations
+    end
+
+    def reservations_by_date(start_date, end_date)
+      s_date = Date.parse(start_date)
+      e_date = Date.parse(end_date)
+      reservations_by_date = []
+      @reservations.each do |reservation|
+        if e_date > reservation.check_in && e_date <= reservation.check_out
+          reservations_by_date << reservation
+        elsif s_date >= reservation.check_in && s_date < reservation.check_out
+          reservations_by_date << reservation
+        elsif s_date <= reservation.check_in && e_date >= reservation.check_out
+          reservations_by_date << reservation
+
+          #Right now checkout day is not included because you can check-in on that day
+        end
+      end
+
+      return reservations_by_date
+    end
+
+    def blocked_rm_by_date(start_date, end_date)
+      s_date = Date.parse(start_date)
+      e_date = Date.parse(end_date)
+      blocks_by_date = []
+      @block.each do |room|
+        if e_date > room.check_in && e_date <= room.check_out
+          blocks_by_date << room
+        elsif s_date >= room.check_in && s_date < room.check_out
+          blocks_by_date << room
+        elsif s_date <= room.check_in && e_date >= room.check_out
+          blocks_by_date << room
+
+          #Right now checkout day is not included because you can check-in on that day
+        end
+      end
+
+      return blocks_by_date
+    end
+
+    def available_rooms(start_date, end_date)
+      unavail_rooms = reservations_by_date(start_date, end_date).map do |reservation|
+        reservation.room_number
+      end
+
+      unavail_block_rms = blocked_rm_by_date(start_date, end_date).map do |block_room|
+        block_room.room_number
+      end
+
+      all_unavail_rooms = unavail_rooms + unavail_block_rms
+
+      all_avail_rooms = all_rooms.reject { |rm_num| all_unavail_rooms.include? rm_num }
```

Comment:
Wonderful wonderful wonderful solution and implementation for this entire piece of functionality! Clean, and very very very clever!